records.go

```diff
@@ -81,29 +116,30 @@ func (dht *IpfsDHT) getPublicKeyFromNode(ctx context.Context, p peer.ID) (ci.Pub
 	// node doesn't have key :(
 	record := pmes.GetRecord()
 	if record == nil {
-		return nil, fmt.Errorf("Node not responding with its public key: %s", p)
+		return nil, fmt.Errorf("Node %v not responding with its public key", p)
 	}
 
-	// Success! We were given the value. we don't need to check
-	// validity because a) we can't. b) we know the hash of the
-	// key we're looking for.
 	val := record.GetValue()
-	log.Debug("DHT got a value from other peer")
+	log.Debugf("DHT got public key from node %v itself", p)
+	return dht.pubkFromVal(p, val)
+}
 
-	pk, err = ci.UnmarshalPublicKey(val)
+func (dht *IpfsDHT) pubkFromVal(p peer.ID, val []byte) (ci.PubKey, error) {
+	pubk, err := ci.UnmarshalPublicKey(val)
 	if err != nil {
+		log.Errorf("DHT could not unmarshall public key for %v", p)
 		return nil, err
 	}
 
-	id, err := peer.IDFromPublicKey(pk)
+	// Make sure the public key matches the peer ID
+	id, err := peer.IDFromPublicKey(pubk)
```

Comment:
Doesn't the record validator already do this? It would be nice to avoid hashing again (although we can probably do that in a future PR).

Reply:
I moved this code so that the public key is only verified for `getPublicKeyFromNode()`, not for `getPublicKeyFromDHT()`. Note that we do need to verify the value we retrieve directly from the node because the node could spoof someone else's key.

Reply:
Ah. I assumed that `getValueSingle` would validate.